### PR TITLE
Bounded state parts

### PR DESCRIPTION
**Pull request by mikhailOK**
Improve state parts splitting logic: now we can have an upper bound
on a state part overhead that depends on max key length and max value
length.

### CODE REVIEW

**Comment by Kouprin** [comment]
What is the overhead? What is expected size of the part after calling `get_trie_nodes_for_part()`?

**Comment by mikhailOK** [comment]
it's going to be  <= `(root_memory_usage/num_parts) + key_length_limit * const1 + value_length_limit * const2`

**Comment by Kouprin** [comment]
What are `key_length_limit` and `value_length_limit`? How they are limited for now? How does it related to transferring via network parts of a state?

**Comment by mikhailOK** [comment]
Right now they are unlimited, they should be limited by our config.

**Comment by Kouprin** [comment]
I expect `key_length_limit` and `value_length_limit` are related to the inner way of storing the data and network layer shouldn't have any ideas about what's going on there. The only thing that network should know about parts - they are bounded to some limit. That's it. For any inner things keep inner constants which should be invisible.

**Comment by mikhailOK** [comment]
part size overhead limit can be a config option too, will need to be consistent with what we decide for storage limits

**Comment by Kouprin** [comment]
Please make sure all the relevant constants are included to this PR or to a separate task.

File: core/store/src/trie/mod.rs [hunk 1]
```diff
@@ -50,7 +50,7 @@ pub struct TrieCosts {
     pub node_cost: u64,
 }
 
-const TRIE_COSTS: TrieCosts = TrieCosts { byte_of_key: 2, byte_of_value: 1, node_cost: 40 };
+const TRIE_COSTS: TrieCosts = TrieCosts { byte_of_key: 2, byte_of_value: 1, node_cost: 50 };
```
Comment (bowenwang1996) [inline core/store/src/trie/mod.rs]: shouldn't this be in some config?

File: core/store/src/trie/mod.rs [hunk 2]
```diff
@@ -575,6 +583,10 @@ impl Trie {
             NodeHandle::InMemory(h) => memory.node_ref(h).clone(),
             NodeHandle::Hash(h) => self.retrieve_node(&h).expect("storage failure"),
         };
+        // Hack so it doesn't slow down some tests
+        if memory_usage > 10_000 {
```
Comment (bowenwang1996) [inline core/store/src/trie/mod.rs]: This should not exist

File: core/store/src/trie/mod.rs (hunk 1 above)
Comment (mikhailOK) [reply core/store/src/trie/mod.rs]: #1744

**Pull request merged**